### PR TITLE
Various API extensions for the fight modules

### PR DESCRIPTION
**Pull request by Vassyli**
Adds various API extensions to make the module flow easier:

- Adds a possibility to expand upon a viewpoint description. So far, only rewrite has been possible.
- Adds more tools to manager actions and action groups in viewpoints (`addActionGroup` to add a group, `findActionGroupById` to return a specific group, `hasActionGroup` to check if a group exists).
- Adds a integer dice to the dice bag
- Adds serialization interface to the Battle object
- Adds a two more specific EventData classes (one for events based on a character, one for events specifically to modify a viewpoint).

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -2,6 +2,7 @@ sudo: false
 language: php
 php:
   - '7.1'
+  - '7.2'
```
Comment (austenmc) [inline .travis.yml:5]: #onward

File: src/Events/CharacterEventData.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
+<?php
+/**
+ * Created by PhpStorm.
+ * User: Basilius Sauter
+ * Date: 04.01.2018
+ * Time: 12:58
+ */
```
Comment (austenmc) [inline src/Events/CharacterEventData.php]: Let's kill this kind of comment. It doesn't add value. I don't care if you'd like to mark it w/ your name, but the IDE+date/time don't seem usefule.

File: src/Events/CharacterEventData.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    /**
+     * CharacterEventData constructor.
+     * @param array $data Must contain field character.
+     * @throws ArgumentException
+     */
+    protected function __construct(array $data)
+    {
+        $mustHaveForm = ["character"];
+        $doesHaveForm = array_keys($data);
+        sort($mustHaveForm); sort($doesHaveForm);
```
Comment (austenmc) [inline src/Events/CharacterEventData.php]: nit: put these on separate lines

File: src/Events/CharacterEventData.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+use LotGD\Core\Models\Character;
+
+/**
+ * Class CharacterEventData
+ * @package LotGD\Core\Events
+ */
+class CharacterEventData extends EventContextData
+{
+    /**
+     * CharacterEventData constructor.
```
Comment (austenmc) [inline src/Events/CharacterEventData.php:16]: Can you expand on what this class is?

File: src/Events/ViewpointDecorationEventData.php [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    /**
+     * ViewpointDecorationEventData constructor.
+     * @param array $data Must contain field viewpoint.
+     * @throws ArgumentException
+     */
+    protected function __construct(array $data)
+    {
+        $mustHaveForm = ["viewpoint"];
+        $doesHaveForm = array_keys($data);
+        sort($mustHaveForm); sort($doesHaveForm);
```
Comment (austenmc) [inline src/Events/ViewpointDecorationEventData.php]: nit: same

File: src/Models/Viewpoint.php [hunk 6]
```diff
@@ -155,6 +230,23 @@ public function findActionGroupById(string $actionGroupId)
         return null;
     }
 
+    /**
+     * Checks if the viewpoint has a certain action group.
+     * @param string $actionGroupId
+     * @return bool
+     */
+    public function hasActionGroup(string $actionGroupId): bool
```
Comment (austenmc) [inline src/Models/Viewpoint.php]: Do we really need this and findActionGroupById? I dont think so.

And if we do, let's just use findActionGroupById() !=== null as the implementation.

File: src/Tools/Model/PropertyManager.php [hunk 7]
```diff
@@ -43,7 +56,12 @@ public function unsetProperty(string $name)
         }
     }
 
-    public function setProperty(string $name, $value)
+    /**
+     * Sets a property to a given value
```
Comment (austenmc) [inline src/Tools/Model/PropertyManager.php:60]: Thanks for adding these docblocks.

image

**Review by austenmc — APPROVED** [review]
Approved but address some of the comments 👍

File: src/Events/CharacterEventData.php (hunk 2 above)
Comment (Vassyli) [reply src/Events/CharacterEventData.php]: I must figure out how to deactivate these comments, they get added from PHPStorm automatically ;) Will delete.

File: src/Models/Viewpoint.php (hunk 6 above)
Comment (Vassyli) [reply src/Models/Viewpoint.php]: I suppose renaming the find-Method to a getMethod would be better. Having both makes the code a bit easier to read:

```php
if ($v->hasActionGroup($actionGroupId)) {
  $v->getActionGroup($actionGroupId)->addAction($action);
} else {
  $v->addActionGroup(
    (new ActionGroup($actionGroupId))
    ->addAction($action)
  );
}
```

vs

```php
$group = $v->getActionGroup($actionGroupId);
if ($group) {
  $group->addAction($action);
} else {
  $v->addActionGroup(
    (new ActionGroup($actionGroupId))
    ->addAction($action)
  );
}
```

File: src/Tools/Model/PropertyManager.php (hunk 7 above)
Comment (Vassyli) [reply src/Tools/Model/PropertyManager.php:60]: They are really valuable, I know. Who would have known that setProperty sets a property!

File: src/Events/CharacterEventData.php (hunk 4 above)
Comment (Vassyli) [reply src/Events/CharacterEventData.php:16]: I plan to make a lot of events that are Character related and do not have any value present except the character (killed-dragon, level-up). Based on this, I plan to introduce a CharacterValueEventData that has a value that needs adjustment, for example in $character->getAttack().

File: src/Tools/Model/PropertyManager.php (hunk 7 above)
Comment (austenmc) [reply src/Tools/Model/PropertyManager.php:60]: Well, it avoids lint spew :)

File: src/Models/Viewpoint.php (hunk 6 above)
Comment (austenmc) [reply src/Models/Viewpoint.php]: Im indifferent to whether it exists. Go with your gut here. But dont write two implementations :)

File: src/Events/CharacterEventData.php (hunk 4 above)
Comment (austenmc) [reply src/Events/CharacterEventData.php:16]: Yeah, we should probably build some kind of codegen for these since they all look the same. #eventually

**Pull request merged**